sklearn/tree/tree.py

```diff
@@ -64,6 +66,119 @@
 SPARSE_SPLITTERS = {"best": _splitter.BestSparseSplitter,
                     "random": _splitter.RandomSparseSplitter}
 
+
+# =============================================================================
+# Support functions
+# =============================================================================
+def preproc_categorical(X, categorical, check_input):
```

Comment:
> Well, I can't make these functions private because they are used by ensembles (forest, gradient boosting).

You can import private functions (whose name is prefixed by underscore) from other modules too... The reason why we we do so is to not expose them to the public API so we don't have to worry about deprecating it when we make modifications to it...

> labels that were not present when fitting the encoder
Ah! There seems to be an effort at #3599 for this but hasn't been merged yet...

It sounds like a good idea to support labels that were not present during fit. Thanks! That will quite often be the case when running cross validation for small datasets.

I am wondering if we should make the `categorical` a dict param taking in the possible categories for each feature? If this sounds like more work for the user we could make it optional? (Either specify a list of int to learn the categories from data or a dict mapping feature id to list of categories to initialize a `LabelEncoder` on it...)

I am suggesting that because once we introduce `partial_fit`, it would be undesirable to put all the labels that were not seen in the first fit to a single higher label...

> After all, it's not the job of decision trees to preprocess the data. Removing this functionality would greatly simplify the changes in this PR and would reduce surface area for bugs.

Ah I'm afraid it is the DT's work to do so as this is not exactly preprocessing but supporting the most common way the categorical values are presented in data which is as string labels...

Maybe @glouppe or @jmschrei has better intuition on what to do in these situations?